### PR TITLE
Port X.509 test parser to Rust

### PR DESCRIPTION
**Pull request by alex**

### CODE REVIEW

File: tests/x509/test_x509.py [hunk 1]
```diff
@@ -1971,9 +1902,13 @@ def test_extreme_times(self, not_valid_before, not_valid_after, backend):
...
         assert cert.not_valid_before == not_valid_before
         assert cert.not_valid_after == not_valid_after
-        parsed = _parse_cert(cert.public_bytes(serialization.Encoding.DER))
-        assert parsed.not_before_tag == UTC_TIME
-        assert parsed.not_after_tag == GENERALIZED_TIME
+        parsed = asn1.test_parse_certificate(
+            cert.public_bytes(serialization.Encoding.DER)
+        )
+        # UTC TIME
+        assert parsed.not_before_tag == 0x17
```
Comment (reaperhulk) [inline tests/x509/test_x509.py:1909]: Should the rust side expose constants for these tags?

File: src/rust/src/asn1.rs [hunk 2]
```diff
@@ -179,6 +179,93 @@ fn encode_dss_signature(
...
+                        let not_before_tag = p.read_element::<asn1::Tlv>()?.tag();
+                        let not_after_tag = p.read_element::<asn1::Tlv>()?.tag();
+                        Ok((not_before_tag, not_after_tag))
+                    })?;
+                    // Subject
+                    let subject_value_tags = parse_name_value_tags(p)?;
+
+                    // Subject public key info
+                    p.read_element::<asn1::Sequence>()?;
+                    // Issuer unique ID
```
Comment (reaperhulk) [inline src/rust/src/asn1.rs:244]: Given what you learned here it's probably worth putting a comment that says "these are optional but literally do not exist in the real world".

File: tests/x509/test_x509.py (hunk 1 above)
Comment (alex) [reply tests/x509/test_x509.py:1909]: Don't really have an opinion on that.

File: src/rust/src/asn1.rs (hunk 2 above)
Comment (alex) [reply src/rust/src/asn1.rs:244]: Sure, I can add a comment.

File: tests/x509/test_x509.py (hunk 1 above)
Comment (reaperhulk) [reply tests/x509/test_x509.py:1909]: I'm comfortable with saying we can revisit this as we progress. It's not like this is a breaking API change so we can add it at any time.

**Review by reaperhulk — APPROVED** [review]

**Pull request merged**